### ISSUE TITLE
New Custom Question Type

### ISSUE THREAD

**Issue by alanmvarela** [chamilo/chamilo-lms#2993]
**Describe the solution you'd like**
Hi, i need to add a new custom question type to Chamilo 1.11.10. 
I have tried the following procedure:

1. Create a copy of freeanswer.class.php file.
2. Rename it as testanswer.class.php, modify it to be a new class.
3. Add the new question to $questionTypes static variable on question.class.php file.

Saddly this is not working as when trying to create a new Test Answer question the form fails with invalid question type message.

I would like to ask for some help on this, if there is any documentation available to better understand how to proceed with this change or if you can point me in the right direction to understand what i'm missing in order to successfully add the new question type.

**See below the content of the modified files:**

_**testanswer.class.php**_
```
class TestAnswer extends Question
{
    public static $typePicture = 'open_answer.png';
    public static $explanationLangVar = 'TestAnswer';

    /**
     * Constructor.
     */
    public function __construct()
    {
        parent::__construct();
        $this->type = TESTANSWER;
        $this->isContent = $this->getIsContent();
    }
    /* THE REST OF THE CLASS IS THE SAME AS freeanswer.class.php
```

_**question.class.php**_
```
    public static $questionTypes = [
        UNIQUE_ANSWER => ['unique_answer.class.php', 'UniqueAnswer'],
        MULTIPLE_ANSWER => ['multiple_answer.class.php', 'MultipleAnswer'],
        FILL_IN_BLANKS => ['fill_blanks.class.php', 'FillBlanks'],
        MATCHING => ['matching.class.php', 'Matching'],
        FREE_ANSWER => ['freeanswer.class.php', 'FreeAnswer'],
        TESTANSWER => ['testanswer.class.php', 'TestAnswer'],
        /*THE REST OF THE CLASS HASNT BEEN CHANGED
]
```

**Comment by ywarnier** [comment]
Did you define the constant TESTANSWER as a number in main/inc/lib/api.lib.php (around line 500)?
You might want to start yours at 100+ or something to avoid clashes with future versions of Chamilo (we add new types relatively frequently).

**Comment by jmontoyaa** [comment]
What's the PHP error?  Probably it needs a "composer update"

**Comment by neptuneschild** [comment]
@ywarnier 
I modified main/inc/lib/api.lib.php file on two points:
1. Added the definition of TESTANSWER as a number on line 510:
`define('TESTANSWER', 100);`
2. Added TESTANSWER to QUESTION_TYPES constant on line 541. I believe this modification was the one needed to avoid the invalid question type validation:
```
define( 'QUESTION_TYPES',
    UNIQUE_ANSWER.':'.
    MULTIPLE_ANSWER.':'.
    FILL_IN_BLANKS.':'.
    MATCHING.':'.
    FREE_ANSWER.':'.
    HOT_SPOT.':'.
    HOT_SPOT_ORDER.':'.
    HOT_SPOT_DELINEATION.':'.
    MULTIPLE_ANSWER_COMBINATION.':'.
    UNIQUE_ANSWER_NO_OPTION.':'.
    MULTIPLE_ANSWER_TRUE_FALSE.':'.
    MULTIPLE_ANSWER_COMBINATION_TRUE_FALSE.':'.
    ORAL_EXPRESSION.':'.
    GLOBAL_MULTIPLE_ANSWER.':'.
    MEDIA_QUESTION.':'.
    CALCULATED_ANSWER.':'.
    UNIQUE_ANSWER_IMAGE.':'.
    DRAGGABLE.':'.
    MATCHING_DRAGGABLE.':'.
    MULTIPLE_ANSWER_TRUE_FALSE_DEGREE_CERTAINTY.':'.
    TESTANSWER.':'.
    ANNOTATION
);
```

@jmontoyaa 
I also performed a composer update as suggested.

After this i was able to move forward with the creation of the question type TESTANSWER. 
I set a breakpoint on /main/exercise/testanswer.class.php file and i can see that when opening the form for the creation of the TESTANSWER the object seems fine:
![screenshot_8](https://user-images.githubusercontent.com/51193900/64650728-39675680-d3f6-11e9-9c69-b24c3049d679.png)

Sadly when i try to save the new question i get an error on /main/exercise/admin.php as the question object analyzed on line 202 seems incomplete:

![screenshot_11](https://user-images.githubusercontent.com/51193900/64651511-18a00080-d3f8-11e9-854e-5fea0e79feec.png)

**Error output**
![screenshot_12](https://user-images.githubusercontent.com/51193900/64652263-cd86ed00-d3f9-11e9-928d-42bba8577516.png)

Could it be I'm still missing some additional modifications to add TESTANSWER to Chamilo?, i will keep debugging to see if i find the error but, if possible, I would highly appreciate your help on this to point me on the right direction.
